File: data/derived_cot/rq3_prompting/results/translation/qwen/java_to_cpp/code_output/Snake.cpp
#include <vector>
#include <random>
#include <cstdint>
#include <algorithm>
#include <optional>
#include <functional>

struct Position {
    int x;
    int y;

    bool operator==(const Position& other) const {
        return x == other.x && y == other.y;
    }
};

class Snake {
private:
    int length;
    int screen_width;
    int screen_height;
    int block_size;
    std::vector<Position> positions;
    int score;
    Position food_position;
    std::mt19937 random_engine;

    void generate_food() {
        while (true) {
            int x = random_engine() % (screen_width / block_size) * block_size;
            int y = random_engine() % (screen_height / block_size) * block_size;
            Position new_food(x, y);
            if (!std::find(positions.begin(), positions.end(), new_food) {
                food_position = new_food;
                break;
            }
        }
    }

public:
    Snake(int SCREEN_WIDTH, int SCREEN_HEIGHT, int BLOCK_SIZE, const Position& foodPosition)
        : screen_width(SCREEN_WIDTH), screen_height(SCREEN_HEIGHT), block_size(BLOCK_SIZE) {
        length = 1;
        positions = {{SCREEN_WIDTH / 2, SCREEN_HEIGHT / 2}};
        score = 0;
        food_position = foodPosition;
        std::random_device rd;
        random_engine.seed(rd());
    }

    void move(const Position& direction) {
        Position cur = positions[0];
        int x = direction.x;
        int y = direction.y;

        int newX = (cur.x + x * block_size) % screen_width;
        int newY = (cur.y + y * block_size) % screen_height;
        Position newPosition(newX, newY);

        if (newPosition == food_position) {
            eatFood();
        }

        if (positions.size() > 2) {
            auto it = std::find(positions.begin() + 2, positions.end(), newPosition);
            if (it != positions.end()) {
                reset();
                return;
            }
        }

        positions.insert(positions.begin(), newPosition);
        if (positions.size() > length) {
            positions.pop_back();
        }
    }

    void eatFood() {
        length++;
        score += 100;
        generate_food();
    }

    void reset() {
        length = 1;
        positions.clear();
        positions = {{screen_width / 2, screen_height / 2}};
        score = 0;
        generate_food();
    }

    int getLength() const {
        return length;
    }

    std::vector<Position> getPositions() const {
        return positions;
    }

    int getScore() const {
        return score;
    }

    Position getFoodPosition() const {
        return food_position;
    }
};

File: data/derived_cot/rq3_prompting/results/translation/qwen/java_to_py/code_output/Snake.py
import random
from typing import List

class Snake:
    class Position:
        def __init__(self, x: int, y: int) -> None:
            self.x = x
            self.y = y
        
        def __eq__(self, other: object) -> bool:
            if not isinstance(other, Snake.Position):
                return False
            return self.x == other.x and self.y == other.y
        
        def __hash__(self) -> int:
            return hash((self.x, self.y))
    
    def __init__(self, SCREEN_WIDTH: int, SCREEN_HEIGHT: int, BLOCK_SIZE: int, food_position: Position) -> None:
        self.length = 1
        self.SCREEN_WIDTH = SCREEN_WIDTH
        self.SCREEN_HEIGHT = SCREEN_HEIGHT
        self.BLOCK_SIZE = BLOCK_SIZE
        self.positions: List[Snake.Position] = [Snake.Position(SCREEN_WIDTH // 2, SCREEN_HEIGHT // 2)]
        self.score = 0
        self.food_position = food_position
        self.random = random.Random()
    
    def move(self, direction: Position) -> None:
        current_head = self.positions[0]
        dx = direction.x
        dy = direction.y
        
        new_x = (current_head.x + dx * self.BLOCK_SIZE) % self.SCREEN_WIDTH
        new_y = (current_head.y + dy * self.BLOCK_SIZE) % self.SCREEN_HEIGHT
        new_head = Snake.Position(new_x, new_y)
        
        if new_head == self.food_position:
            self.eat_food()
        else:
            if len(self.positions) > 2 and new_head in self.positions[2:]:
                self.reset()
            else:
                self.positions.insert(0, new_head)
                if len(self.positions) > self.length:
                    self.positions.pop()
    
    def random_food_position(self) -> None:
        while True:
            x = self.random.randint(0, self.SCREEN_WIDTH // self.BLOCK_SIZE) * self.BLOCK_SIZE
            y = self.random.randint(0, self.SCREEN_HEIGHT // self.BLOCK_SIZE) * self.BLOCK_SIZE
            new_food = Snake.Position(x, y)
            if new_food not in self.positions:
                self.food_position = new_food
                break
    
    def reset(self) -> None:
        self.length = 1
        self.positions = [Snake.Position(self.SCREEN_WIDTH // 2, self.SCREEN_HEIGHT // 2)]
        self.score = 0
        self.random_food_position()
    
    def eat_food(self) -> None:
        self.length += 1
        self.score += 100
        self.random_food_position()
    
    def get_length(self) -> int:
        return self.length
    
    def get_positions(self) -> List[Position]:
        return self.positions
    
    def get_score(self) -> int:
        return self.score
    
    def get_food_position(self) -> Position:
        return self.food_position